### ISSUE TITLE
Provide the DownloadAction as a Gradle service

### ISSUE THREAD

**Issue by DenWav** [michel-kraemer/gradle-download-task#161]
This has 2 major use cases:

 * When using the Worker API it's possible for services to be injected into the workers, but you can't get `Project`s injected into workers, so it's not possible (without ugly hacks) to use the `DownloadAction` standalone inside of a worker.
 * With the new configuration cache feature coming in Gradle 6.7 tasks which call `getProject()` for any reason will not allow a build to use that feature, as described [here](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution)

Further explanation on _why_ `Project` isn't allowed to be used is also in there, but one of the things it allows is for all tasks to execute in parallel by default. Since `Project` is not thread-safe, it can't be used in parallel. In some ways the configuration cache feature closes the gap between Tasks and Workers.

 Looking at the code there's currently no way to create a `DownloadAction` without a `Project`.

**Comment by DemonWav** [comment]
I've looked into this a bit more and the biggest issue I see here, if I'm not mistaken, is the `ProgressLogger`. I do not understand why this  is an internal API while the `Worker` API is public and `Tasks` can now execute in parallel, this generally makes no sense to me. This plugin relies on using that to publish the download status like internal Gradle does for configuration resolution, but as far as I can tell you _must_ have a `Project` instance to retrieve that, since Gradle won't give you one through dependency injection.

That seems like a massive oversight from the Gradle team.

**Comment by michel-kraemer** [comment]
I don't think custom build services are available for injection. The following page says that build service properties have to be marked "as @Internal. Using a service with any other annotation is currently not supported" (https://docs.gradle.org/current/userguide/build_services.html). The section on using custom build services shows that a provider must be used and the property must be set manually. Also, the page on developing custom gradle types only lists a few services that can be injected (https://docs.gradle.org/current/userguide/custom_gradle_types.html#service_injection).

To be sure about this, I tested it with the following build script:

```gradle
import org.gradle.api.provider.Property
import org.gradle.api.services.BuildService
import org.gradle.api.services.BuildServiceParameters

public abstract class MyService implements BuildService<MyService.Params> {
    interface Params extends BuildServiceParameters {
    }

    def run() {
      println("Hello")
    }
}

public abstract class MyTask extends DefaultTask {
    @Inject
    abstract Property<MyService> getMyService()

    @TaskAction
    public void run() {
        getMyService().get().run()
    }
}

def s = project.getGradle().getSharedServices().registerIfAbsent("myService", MyService.class, spec -> {})

task myTask(type: MyTask)
```

If you execute this build script, you will get the following exception:

```
> Could not create task ':myTask'.
   > Could not create task of type 'MyTask'.
      > Could not generate a decorated class for type MyTask.
         > Cannot use @Inject annotation on method MyTask.getMyService().
```

Instead, what you need to do is to change `@Inject` to `@Internal` and then define the task as follows:

```gradle
task myTask(type: MyTask) {
  getMyService().set(s)
}
```

This is similar to what is described in the documentation and works properly.

I'm closing this issue as invalid. If you have feedback or different requirements, feel free to comment or open up a new issue.
